primer/src/Primer/Def/Utils.hs

```diff
@@ -27,3 +31,32 @@ globalInUse v =
   anyOf
     (folded % #_DefAST % #astDefExpr % to freeGlobalVars)
     (Set.member v)
+
+-- | Is this type (including any of its constructors) in use in the given definitions?
+typeInUse :: (Foldable f, Foldable g, Data b, Ord b) => TyConName -> ASTTypeDef a -> f (TypeDef b) -> g Def -> Bool
+typeInUse defName def ts ds =
+  anyOf
+    (folded % #_TypeDefAST % to tyConsInTypeDef)
+    (Set.member defName . Set.map fst)
+    ts
+    || anyOf
+      (folded % #_DefAST % to tyConsInDef)
+      (Set.member defName . Set.map fst)
+      ds
+    || anyOf
+      (folded % #_DefAST % to valConsInDef)
+      (\s -> any ((`elem` Set.map fst s) . valConName) $ astTypeDefConstructors def)
+      ds
+  where
+    tyConsInExpr =
+      Set.unions . map tyConsInType . concatMap (toListOf typesInExpr) . universe
+    tyConsInType t =
+      Set.fromList [(n, m) | TCon m n <- universe t]
+    tyConsInDef d =
+      tyConsInExpr (astDefExpr d) `Set.union` tyConsInType (astDefType d)
+    tyConsInTypeDef =
+      Set.unions . map (Set.unions . map tyConsInType . valConArgs) . astTypeDefConstructors
+    valConsInExpr e =
+      Set.fromList [(n, m) | Con m n _ <- universe e]
+    valConsInDef =
+      valConsInExpr . astDefExpr
```

Comment:
We could inline some of these helpers and avoid creating sets at all. Originally I was thinking I'd need these elsewhere, but in the end I only needed to export `typeInUse`. Just traversing instead should even take complexity down from O(n log n) to O(n), although it's possible that rewrites+fusion save us anyway in practice.